tests/core/test_metric_result_integration.py

```diff
@@ -175,3 +184,109 @@ def lightning_log(fx, *args, **kwargs):
         for k in ('d0.a', 'd1.a'):
             assert result[k].value == torch.tensor(3.) + epoch, k
             assert result[k].cumulated_batch_size == torch.tensor(1.), k
+
+
+def test_result_collection_restoration():
+    """"
+    This test make sure metrics are properly reloaded on failure.
+    """
+
+    result = ResultCollection(True, torch.device("cpu"))
+    _result = None
+    metric_a = DummyMetric('a')
+    metric_b = DummyMetric('b')
+    metric_c = DummyMetric('c')
+    metric_d = DummyMetric('d')
+    current_fx_name = None
+    batch_idx = None
+
+    def lightning_log(fx, *args, **kwargs):
+        nonlocal current_fx_name
+        if current_fx_name != fx and batch_idx in (None, 0):
+            result.reset(metrics=False, fx=fx)
+        result.log(fx, *args, **kwargs)
+        current_fx_name = fx
+
+    for _ in range(2):
+
+        cumulative_sum = 0
+
+        for i in range(3):
+
+            a = metric_a(i)
+            b = metric_b(i)
+            c = metric_c(i)
+            metric_d(i)
+
+            cumulative_sum += i
+
+            metric = metric_a if i < 1 else metric_d
+            lightning_log('training_step', 'a', metric, on_step=True, on_epoch=True)
+            lightning_log('training_step', 'b', metric_b, on_step=False, on_epoch=True)
+            lightning_log('training_step', 'c', metric_c, on_step=True, on_epoch=False)
+            lightning_log('training_step', 'a_1', a, on_step=True, on_epoch=True)
+            lightning_log('training_step', 'b_1', b, on_step=False, on_epoch=True)
+            lightning_log('training_step', 'c_1', {'1': c, '2': c}, on_step=True, on_epoch=False)
+
+            batch_log = result.metrics(on_step=True)[MetricSource.LOG]
+            assert set(batch_log) == {"a_step", "c", "a_1_step", "c_1"}
+            assert set(batch_log['c_1']) == {'1', '2'}
+            _result = deepcopy(result)
+            state_dict = result.state_dict()
+
+            result = ResultCollection(True, torch.device("cpu"))
+            result.load_state_dict(state_dict, sync_fn=_result['training_step.a'].meta.sync.fn)
+
+            assert _result.items() == result.items()
+            assert _result["training_step.c_1"].meta == result["training_step.c_1"].meta
+
+        batch_idx = None
+
+        epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
+        _epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
```

Comment:
```suggestion
        epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
        _epoch_log = _result.metrics(on_step=False)[MetricSource.LOG]
```
Maybe this was the intention?
Anyway it's not so clear why we test that the outputs of a deepcloned result is the same. 
maybe put a comment what the difference is between _result and result ?

Reply:
Good catch ! Renamed result_copy for clarity.